### PR TITLE
feat: 476 add full site search

### PR DESCRIPTION
**Pull request by hetd54**
Uses [Lunr](https://lunrjs.com/) in order to implement a full-site search

Adds a new layout, SearchPageLayout, and a new page at the "/search" url.

### CODE REVIEW

**Review by hollandjg — CHANGES_REQUESTED** [review]
This feels excellent, Heather! The quick search button is great and the results come quick. 

My main concern is that the Search is now an integral part of the Layout, but I don't think they are really that closely connected – the one is to do with the styling of each page (and is really pretty simple), the other is to do with searching the entirety of the content (and has additional dependencies beyond just Gatsby and our styling). I think it breaks separation of concerns. It also means that you've got to load all the data for the search each time you load any page on the site, which might get expensive.

I'd suggest the search be an independent page for now, which we will aim to incorporate as a modal in a next step. That would mean that the new page has its own query, and the <Layout> and its query doesn't need to be modified here. (I imagine having a modal which loads the search page as a separate thing, maybe in an iframe, though I'm not sure if we really want to use frames.) You could put the code in the `src/pages/` directory so that it's automatically generated for every site, and then we just link it from the NavBar when we want to activate search functionality, like the 404 page.

Would it be possible to extract just the search-results part into an independent layout? The SearchPageLayout doesn't seem to be working right now in the Storybook.

I've got some comments on the styling which may or may not apply to the SearchPageLayout as you have it right now:
- The results get a bit squished on mobile. Would it be possible to add defaults to the tailwind config which work better for mobile (narrower or no margins), and then add the current settings with `m:` or `l:` or `xl:` so that it looks the same on larger screens? https://tailwindcss.com/docs/responsive-design image
- The centering of the result texts feels a bit inconsistent to me, given that the cards use left-aligned text. Could you update the styling so that it's more consistent? image 
- The whole of each card is typically a link to the page, and I would expect the listed title from the search result to be a link. Would it be a good idea to make the whole of the result card into a link? (and maybe remove the "view page" button – I'm not sure that's needed for search results. I'm willing to be persuaded otherwise though.)
<details><summary>Browser metadata</summary>





```
Path:      /
Browser:   Safari 17.0 on Mac OS 10.15.7
Viewport:  1054 x 674 @3x
Language:  en-US
Cookies:   Disabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-653--gatsby-theme-project-portal-ex-site.netlify.app%2F&os=OS%20X&os_version=Catalina&browser=Safari&browser_version=17.0&resolution=1054x674&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-653--gatsby-theme-project-portal-ex-site.netlify.app/) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=65298dd0b2f9ed000809cb03&commentID=6532ce1ba3bdb702bcd2d258&resolution=resolved)

File: package.json [hunk 1]
```diff
@@ -31,6 +31,8 @@
     "node": ">=18.0.0"
   },
   "dependencies": {
-    "gatsby-plugin-testing": "^0.3.5"
+    "gatsby-plugin-testing": "^0.3.5",
+    "lunr": "^2.3.9",
+    "react-modal": "^3.16.1"
```
Comment (hollandjg) [inline package.json:36]: I think this needs to be in packages/gatsby-theme-project-portal/package.json rather than here

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    {db[result.ref].question || db[result.ref].slug}
+                  </h2>
+                  <ul className="grid grid-rows-1 justify-center md:block">
+                    <h2 className="font-sans text-black text-tag mb-1 font-extrabold">
+                      Match found in:
+                    </h2>
+                    {found}
+                  </ul>
+                  <h2 className="pt-3 mt-4 border-t-2 border-solid">
+                    <a className="btn-wide" href={`${siteUrl}/${result.ref}`}>
```
Comment (jashlu) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:165]:
```suggestion
                    <a className="btn-wide" href={`${siteUrl}/project/${result.ref}`}>
```

**Review by jashlu — CHANGES_REQUESTED** [review]

**Comment by jashlu** [comment]
https://lunrjs.com/guides/index_prebuilding.html

Could this be an alternative?

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx (hunk 2 above)
Comment (hetd54) [reply packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:231]: I am adding "project" to each project's slug prior to this, since some pages won't have it (about page, contact page, etc)

File: package.json (hunk 1 above)
Comment (hetd54) [reply package.json]: done!

**Comment by hetd54** [comment]
Notes from meeting:

- We will have a separate PR to implement styling for the results to match the project cards
- We will have a separate PR to put the search feature into a modal
- TODO: make sure the index only gets built once, on build, instead of when the page is mounted/unmounted

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        faq: {
+          text: string
+          title: string
+        }[]
+        aims: {
+          text: string
+          title: string
+        }[]
+        title: string
+      }
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:58]: Doesn't this need to be an array too?
```suggestion
      }[]
```

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        statusOfData: string
+        fundingInfo: string
+        emailContent: string
+        mainContact: ContactType
+        projectTeam?: ContactType[]
+        faq: {
+          text: string
+          title: string
+        }[]
+      }
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:43]: I think this needs to be an array:
```suggestion
      }[]
```

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+export default meta
+
+type Story = StoryObj<typeof SearchPageLayout>
+
+export const Primary: Story = {
+  args: {
+    data: {
+      allGeneralPage: {
+        nodes: {
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx:18]: Do we need the `nodes` here to be an array? I'm getting an undefined error when I run this on the preview https://deploy-preview-653--project-portal-storybook.netlify.app/?path=/docs/gatsby-theme-project-portal-src-layouts-searchpagelayout--docs
I think it's because the nodes isn't an array, and the `foreach` is then undefined.

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+          slug: "/about",
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx:19]: I think you might be showing the slug rather than the page title at the top of the results card for these general pages:
Screenshot 2023-10-26 at 16 34 51

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+          endDate: new Date("2022-03-04"),
+          agency: "Sample Agency",
+          lastModified: new Date("2022-05-27T16:34:04.000Z"),
+          topics: [{ slug: "test", title: "Test" }],
+        },
+      },
+      site: { siteMetadata: { siteUrl: "https://localhost" } },
+    },
+  },
+}
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx:78]: This looks like a place where we could benefit from having interaction tests built-in to storybook. Check this out: https://storybook.js.org/docs/react/writing-stories/play-function

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+        deliverable
+        purpose: string
+        expertise: string
+        requirement: string
+        keyDates: string
+        priorResearch: string
+        statusOfData: string
+        fundingInfo: string
+        emailContent: string
+        mainContact: ContactType
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:37]: are these all required? I think they're likely to be `null | undefined` a lot of the time, so perhaps we shouldn't rely on them being there.

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        slug: string
+        lede: string
+        faq: {
+          text: string
+          title: string
+        }[]
+        aims: {
+          text: string
+          title: string
+        }[]
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:56]: same here – aims only applies to About, not to the contact page, so these fields should be optional

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx:79]: What happens when we have no projects to filter, or just one, or two? What about when we have no general pages to filter, or just one, or two?

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            {
+              title: "Is this an FAQ question?",
+              text: "Yes, and here is the answer.",
+            },
+          ],
+          title: null,
+        },
+      },
+      allProject: {
+        nodes: {
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.stories.tsx:37]: You could consider reusing the data from other stories here. You can do something like this:
```typescript
import * as AboutPageStories from "./AboutPageLayout.stories"

...

export const OneGeneralPage: Story = {
  args: {
    data: {
      allGeneralPage: {
        nodes: [
          {
            slug: "about",
            ...AboutPageStories.Primary.args.data.generalPage,
          },
        ],
      },
      allProject: {
        nodes: [],
      },
      site: { siteMetadata: { siteUrl: "https://localhost" } },
    },
  },
}
```

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+    /* allow stop words (see https://github.com/olivernn/lunr.js/blob/master/lunr.js)
+     we have 'about' as one of the slugs, which spurred this decision
+    We can implement our own stopWordFilter:
+    var idx = lunr(function () {
+      // normal index definition
+      this.pipeline.after(lunr.stopWordFilter, myStopWordFilter)
+       this.pipeline.remove(lunr.stopWordFilter)
+    })
+     */
+    this.pipeline.remove(lunr.stopWordFilter)
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:154]: could we remove "about" from the lunr.stopWordFilter and still have the benefit of the filter? Doesn't look too complicated to reuse: https://lunrjs.com/docs/stop_word_filter.js.html

**Review by hollandjg — COMMENTED** [review]
I found some things you might want to look at in the Storybook config!

**Review by hollandjg — CHANGES_REQUESTED** [review]
Hey Heather! I'm afraid this crashes on me - I get the error message that the `index` is undefined, and that the lunr-index.json doesn't exist.
It only happens the first time I build the site – it looks like the assets you add to the static directory at the end of the gatsby-node step don't necessarily get included in the built site. Second time it runs fine! Unfortunately, that means it doesn't work on Netlify, as we only built the site once on there.
So close!

**Comment by jashlu** [comment]
> Hey Heather! I'm afraid this crashes on me - I get the error message that the `index` is undefined, and that the lunr-index.json doesn't exist. It only happens the first time I build the site – it looks like the assets you add to the static directory at the end of the gatsby-node step don't necessarily get included in the built site. Second time it runs fine! Unfortunately, that means it doesn't work on Netlify, as we only built the site once on there. So close!

Played around with the build this morning and at first I also encountered the same thing. I would build the site and serve it immediately after the build completion, I'm guessing you did the same. 

I decided then to wait a couple minutes after the build to serve and sure enough the two json documents show up in the static folder. I believe, because of the size of the content being written into the files, it takes a while before the two files can be fully written. Of course serving the site right after build when the two static files are not yet fully written will mean the site is served without those two documents. It will be something we have to take into consideration, considering the index will only continue to grow bigger.

**Comment by jashlu** [comment]
> > Hey Heather! I'm afraid this crashes on me - I get the error message that the `index` is undefined, and that the lunr-index.json doesn't exist. It only happens the first time I build the site – it looks like the assets you add to the static directory at the end of the gatsby-node step don't necessarily get included in the built site. Second time it runs fine! Unfortunately, that means it doesn't work on Netlify, as we only built the site once on there. So close!
> 
> Played around with the build this morning and at first I also encountered the same thing. I would build the site and serve it immediately after the build completion, I'm guessing you did the same.
> 
> I decided then to wait a couple minutes after the build to serve and sure enough the two json documents show up in the static folder. I believe, because of the size of the content being written into the files, it takes a while before the two files can be fully written. Of course serving the site right after build when the two static files are not yet fully written will mean the site is served without those two documents. It will be something we have to take into consideration, considering the index will only continue to grow bigger.

Looking into the write function right now, maybe it has something to do with our encoding

File: packages/gatsby-theme-project-portal/gatsby-node.js [hunk 13]
```diff
@@ -129,3 +194,157 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
...
+    return { index, documents }
+  }
+
+  const { index, documents } = createSearchIndex({ allProject, allGeneralPage })
+  fs.writeFile("static/lunr-index.json", JSON.stringify(index), (err) => {
+    if (err) console.error(err)
+  })
+  fs.writeFile("static/documents.json", JSON.stringify(documents), (err) => {
+    if (err) console.error(err)
+  })
```
Comment (jashlu) [inline packages/gatsby-theme-project-portal/gatsby-node.js:345]: I'm not 100% certain but I believe the code we have currently is fine because the exports.onPostBuild section is meant for running asynchronous functions so it will complete the writeFile before serving the site (the async key word on line 198) 

Another option is for us to make the writeFile synchronous rather than asynchronous. 

```
try {
    fs.writeFileSync("static/lunr-index.json", JSON.stringify(index));
    console.log('lunr-index.json written successfully.');

   fs.writeFileSync("static/documents.json", JSON.stringify(documents));
    console.log('documents.json written successfully.');
  } catch (err) {
    console.error('Error writing files:', err);
  }
```

This should guarantee the file write is done before the build process is wrapped up. But it does essentially block the build process until the write operation is complete, which could affect performance overall

Comment (hetd54) [reply packages/gatsby-theme-project-portal/gatsby-node.js:345]: @jashlu onPostBuild is also async for awaiting the graphql query used to create the index -- will making the writeFile async still solve our issue?

Comment (jashlu) [reply packages/gatsby-theme-project-portal/gatsby-node.js:345]: I believe that should do it ..

**Review by jashlu — APPROVED** [review]

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx (hunk 12 above)
Comment (hetd54) [reply packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx]: @hollandjg Are there any other words within the general stop word list that we should re-add?

[
    "a",
    "able",
    "across",
    "after",
    "all",
    "almost",
    "also",
    "am",
    "among",
    "an",
    "and",
    "any",
    "are",
    "as",
    "at",
    "be",
    "because",
    "been",
    "but",
    "by",
    "can",
    "cannot",
    "could",
    "dear",
    "did",
    "do",
    "does",
    "either",
    "else",
    "ever",
    "every",
    "for",
    "from",
    "get",
    "got",
    "had",
    "has",
    "have",
    "he",
    "her",
    "hers",
    "him",
    "his",
    "how",
    "however",
    "i",
    "if",
    "in",
    "into",
    "is",
    "it",
    "its",
    "just",
    "least",
    "let",
    "like",
    "likely",
    "may",
    "me",
    "might",
    "most",
    "must",
    "my",
    "neither",
    "no",
    "nor",
    "not",
    "of",
    "off",
    "often",
    "on",
    "only",
    "or",
    "other",
    "our",
    "own",
    "rather",
    "said",
    "say",
    "says",
    "she",
    "should",
    "since",
    "so",
    "some",
    "than",
    "that",
    "the",
    "their",
    "them",
    "then",
    "there",
    "these",
    "they",
    "this",
    "tis",
    "to",
    "too",
    "twas",
    "us",
    "wants",
    "was",
    "we",
    "were",
    "what",
    "when",
    "where",
    "which",
    "while",
    "who",
    "whom",
    "why",
    "will",
    "with",
    "would",
    "yet",
    "you",
    "your",
  ])

File: __tests__/gatsby-theme-project-portal/SearchBar.spec.tsx [hunk 14]
```diff
@@ -58,15 +83,25 @@ describe("SearchBar", () => {
...
     const { getByLabelText } = render(
-      <SearchBar label="Test Label" onChange={onChange} />
+      <SearchBar
+        label="Test Label"
+        onChange={onChange}
+        id={"search"}
+        placeholder={"Type to filter posts..."}
+      />
     )
     expect(getByLabelText("Search")).toHaveStyle({ height: "62%" })
```
Comment (hollandjg) [inline __tests__/gatsby-theme-project-portal/SearchBar.spec.tsx:106]: Can you say a bit more about this test? I may be misunderstanding it.
It looks like you're checking the styling – I think that would better be handled using a snapshot test, as the existing test is likely to be brittle if we change the styling. If so, I'd remove it here and reintroduce this kind of testing in the PR where we do the modal.

File: packages/gatsby-theme-project-portal/gatsby-node.js [hunk 15]
```diff
@@ -34,8 +34,54 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
...
+          }
+          deliverable
+          emailContent
+          endDate
+          slug
+        }
+      }
+      allGeneralPage {
         nodes {
           slug
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/gatsby-node.js:74]: Looks here like you might have inadvertently removed the slug from the `allCardPage` part of the query.

File: packages/gatsby-theme-project-portal/gatsby-node.js [hunk 16]
```diff
@@ -34,8 +34,54 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
...
+            text
+            title
+          }
+          deliverable
+          emailContent
+          endDate
+          slug
+        }
+      }
+      allGeneralPage {
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/gatsby-node.js:72]: I'm not sure you need this block of added query – all your query data is nicely isolated in the `onPreBuild` function, is it not? Maybe revert these changes.

File: packages/gatsby-theme-project-portal/gatsby-node.js [hunk 17]
```diff
@@ -129,3 +193,78 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
...
+          }
+          aims {
+            text
+            title
+          }
+          title
+        }
+      }
+    }
+  `)
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/gatsby-node.js:252]: Could this be moved to a graphQL fragment in the utils/search.js file? That's where it's actually "used".
This might not be possible right now – the utils isn't in the `src` directory (though why it isn't is a bit beyond me, no idea why I did that) so Gatsby might not recognize fragments there.

File: packages/gatsby-theme-project-portal/package.json [hunk 18]
```diff
@@ -11,6 +11,7 @@
     "Aisulu Omar <aisulu_omar@brown.edu>",
     "Joselynn Wallace <Joselynn Wallace@brown.edu>",
     "Kevin Wilson <kevin_wilson@brown.edu>",
+    "Aidan Hudson-Lapore <aidan_hudson-lapore@brown.edu>",
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/package.json:14]: Duplicate entry
```suggestion
```

File: packages/gatsby-theme-project-portal/src/components/ProjectPage.tsx
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/ProjectPage.tsx:1]: We need to rename this component – it's confusing with the projectDetailPage. Is there a better name we can think of? Perhaps we should just make this into a separate issue.

File: packages/gatsby-theme-project-portal/src/components/SearchBar.tsx
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SearchBar.tsx:1]: I'm not sure if these changes belong in this PR, or in a separate one. Do they have to do with the full site search?

Comment (hollandjg) [reply packages/gatsby-theme-project-portal/src/components/SearchBar.tsx:1]: Is it the same component as you used in the Modal – just generalized? In that case, great!

File: packages/gatsby-theme-project-portal/src/components/SiteSearch.stories.tsx [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+    { ...ProjectDetailStories.Ongoing.args },
+    { ...ProjectDetailStories.Completed.args },
+  ],
+}
+
+export const oneProject = {
+  nodes: [{ ...ProjectDetailStories.LongTitle.args }],
+}
+
+export const twoGenPages = {
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SiteSearch.stories.tsx:33]:
```suggestion
export const twoGeneralPages = {
```

File: packages/gatsby-theme-project-portal/src/components/SiteSearch.stories.tsx
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SiteSearch.stories.tsx:1]: This is great!

File: packages/gatsby-theme-project-portal/src/components/SiteSearch.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+  nodes: [
+    { ...ProjectDetailStories.LongTitle.args },
+    { ...ProjectDetailStories.Minimum.args },
+    { ...ProjectDetailStories.NullValues.args },
+    { ...ProjectDetailStories.Open.args },
+    { ...ProjectDetailStories.Ongoing.args },
+    { ...ProjectDetailStories.Completed.args },
+  ],
+}
+
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SiteSearch.stories.tsx:28]: Can we maybe move the `Primary` story to the top? It showcases all the standard features of this component, and would be the first thing people see when they open up the Storybook.

File: packages/gatsby-theme-project-portal/src/components/SiteSearch.tsx [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+                className="inline mx-px px-3 py-1 mt-1 text-tag whitespace-nowrap font-semibold mr-2 bg-topics"
+              >
+                {field}
+              </li>
+            ))
+            const item = (
+              <React.Fragment key={result.ref}>
+                <div className="px-2 py-4 overflow-hidden bg-white border border-gray-200 rounded-md shadow-md h-full flex flex-col md:mx-10 md:py-6 lg:mx-60">
+                  <h2 className="capitalize">
+                    {db[result.ref].question || db[result.ref].slug}
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SiteSearch.tsx:70]: I'd think about using the page title here in both cases.

File: packages/gatsby-theme-project-portal/src/components/SearchBar.tsx
Comment (hollandjg) [reply packages/gatsby-theme-project-portal/src/components/SearchBar.tsx:1]: I understand. This is used in the SiteSearch. All good, no changes here! Let's resolve this thread.

File: packages/gatsby-theme-project-portal/src/components/SiteSearchWrapper.tsx [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+      setIdx(lunr.Index.load(savedIndex))
+      // this is a function which grabs the slug from the original documents
+      // lunr tosses this info for SPEED
+      setDb(
+        documents.reduce(function (acc, document) {
+          acc[document.slug] = document
+          return acc
+        }, {})
+      )
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SiteSearchWrapper.tsx:31]: Can we do the documents.reduce in advance too, and just load the db?

File: packages/gatsby-theme-project-portal/src/components/SiteSearchWrapper.tsx [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    // Because Gatsby runs this on build, we need to say to ignore getting the json files
+    // if the window is undefined (not running in the browser)
+    // see: https://www.gatsbyjs.com/docs/debugging-html-builds/
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/components/SiteSearchWrapper.tsx:36]: This is a helpful comment, thanks for including it.

File: packages/gatsby-theme-project-portal/src/fragments/site-search.js
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/fragments/site-search.js:1]: Oh, you already have the Fragment! Let's use it.

File: packages/gatsby-theme-project-portal/src/layouts/AboutPageLayout.stories.tsx
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/AboutPageLayout.stories.tsx:1]: Is this extra test-data?

File: packages/gatsby-theme-project-portal/src/layouts/ContactPageLayout.tsx [hunk 24]
```diff
@@ -20,7 +20,7 @@ interface ContactProps {
     generalPage: {
       title: string
       lede: string
-      image: ImageDataLike
+      image?: ImageDataLike
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/ContactPageLayout.tsx:23]: Looks like a change to fix a typing error – correct?

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+export interface SearchLayoutProps {
+  data: {
+    site: {
+      siteMetadata: {
+        siteUrl: string
+      }
+    }
+    generalPage: {
+      title: string
+      image: ImageDataLike
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:16]:
```suggestion
      image?: ImageDataLike
```

File: packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx [hunk 26]
```diff
@@ -0,0 +1,62 @@
+import React, { FunctionComponent } from "react"
+import { graphql } from "gatsby"
+import SiteSearchWrapper from "../components/SiteSearchWrapper"
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/src/layouts/SearchPageLayout.tsx:3]: might want to export the SiteSearchWrapper in the ../components/index.tsx file, so you can merge this with the line below.

File: packages/gatsby-theme-project-portal/utils/search.js [hunk 27]
```diff
@@ -0,0 +1,202 @@
+const lodash = require("lodash")
+const lunr = require("lunr")
+function createSearchIndex(searchNodes) {
+  // searchNodes = {allProject[], allGeneralPage[], siteUrl}
+  let documents = []
+  Object.keys(searchNodes).forEach((page) => {
+    if (page === "siteUrl") {
+      return
```
Comment (hollandjg) [inline packages/gatsby-theme-project-portal/utils/search.js:8]: What's going on here? I don't understand this line.

Comment (hollandjg) [reply packages/gatsby-theme-project-portal/utils/search.js:8]: Looks like you actually want 
```
function createSearchIndex(searchNodes, siteUrl)
```
at the top of the function

Comment (hollandjg) [reply packages/gatsby-theme-project-portal/utils/search.js:8]: What's the siteUrl used for, anyway? Can you not pass it and get the same effect?

**Review by hollandjg — CHANGES_REQUESTED** [review]
Hey Heather! Thanks for all this work! I've taken a look at everything, and it looks good. There are a couple of places where I think there are still bits and pieces left over from the coding process – in particular the definition of the search data in graphQL. If you can deduplicate those bits, I think that's the last piece. 
(I've made some other questions and comments too, mainly so I understand better what's going on, but that was the main one.)

File: __tests__/gatsby-theme-project-portal/SearchBar.spec.tsx (hunk 14 above)
Comment (hetd54) [reply __tests__/gatsby-theme-project-portal/SearchBar.spec.tsx:106]: Heya, so I updated this test because I changed the props required for the searchbar component. The test will produce an error since all the required props aren't there. The id really needs to be a required prop since that determines focus.

As for the test itself, that was created in a different pr; do we want to look over our tests in another PR?

File: packages/gatsby-theme-project-portal/gatsby-node.js (hunk 17 above)
Comment (hetd54) [reply packages/gatsby-theme-project-portal/gatsby-node.js:252]: A fragment doesn't work, but what about adding the query in the module's exports?

```
const searchQuery = `
    query {
      allProject {
        nodes {
          title
          agency
          topics {
            title
          }
          slug
          summary
          statusOfData
          status
          startDate
          requirement
          question
          purpose
          projectTeam {
            name
            employer
          }
          priorResearch
          opportunityCloses
          mainContact {
            name
          }
          fundingInfo
          expertise
          faq {
            text
            title
          }
          deliverable
          emailContent
          endDate
          slug
        }
      }
      allGeneralPage {
        nodes {
          slug
          lede
          faq {
            text
            title
          }
          aims {
            text
            title
          }
          title
        }
      }
    }
  `

module.exports = { createSearchIndex, searchQuery }
```

**Review by hollandjg — APPROVED** [review]
🚀 Thanks Heather! Looks good to me. 👍

File: packages/gatsby-theme-project-portal/src/components/ProjectPage.tsx
Comment (jashlu) [reply packages/gatsby-theme-project-portal/src/components/ProjectPage.tsx:1]: maybe `projectsDisplayPage`? or `projectsHomePage`?

New issue here https://github.com/thepolicylab-projectportals/project-portal-theme/issues/709
